pkg/storegateway/bucket_index_reader.go

```diff
@@ -268,6 +224,98 @@ func (r *bucketIndexReader) expandedPostings(ctx context.Context, ms []*labels.M
 	return ps, nil
 }
 
+// toPostingGroups returns a set of labels for which to look up postings lists. It guarantees that
+// each postingGroup's keys exist in the index. The order of the returned labels
+// is the same as iterating through the posting groups and for each group adding all keys.
+func toPostingGroups(ms []*labels.Matcher, indexhdr indexheader.Reader) ([]postingGroup, []labels.Label, error) {
+	var (
+		rawPostingGroups = make([]rawPostingGroup, 0, len(ms))
+		allRequested     = false
+		hasAdds          = false
+	)
+
+	// NOTE: Derived from tsdb.PostingsForMatchers.
+	for _, m := range ms {
+		// Each group is separate to tell later what postings are intersecting with what.
+		rawPostingGroups = append(rawPostingGroups, toRawPostingGroup(m))
+	}
+
+	// We can check whether their requested values exist
+	// in the index. We do these checks in a specific order, so we minimize the reads from disk
+	// and/or to minimize the number of label values we help in memory (assumption being that indexhdr.LabelValues
+	// is likely to return a lot of values).
+	sort.Slice(rawPostingGroups, func(i, j int) bool {
+		// First we check the non-lazy groups, since for those we only need to call PostingsOffset, which
+		// is less expensive than LabelValues for the lazy groups.
+		ri, rj := rawPostingGroups[i], rawPostingGroups[j]
+		if ri.isLazy != rj.isLazy {
+			return !ri.isLazy
+		}
+
+		// Within the lazy/non-lazy groups we sort by the number of keys they have.
+		// The idea is that groups with fewer keys are more likely to match no actual series.
+		if len(ri.keys) != len(rj.keys) {
+			return len(ri.keys) < len(rj.keys)
+		}
+
+		// Sort by label name to make this a deterministic sort.
+		return ri.labelName < rj.labelName
+	})
+
+	filteredPostingGroups := make([]postingGroup, 0, len(rawPostingGroups))
```

Comment:
[nit] For the eventual "all postings" we may add. I would also add a comment to explain why +1.

```suggestion
	filteredPostingGroups := make([]postingGroup, 0, len(rawPostingGroups) + 1)
```

Reply:
All postings is a _key_, not a _group_, and we account for it later in `numKeys`.

Reply:
But we add it to the groups too, no?
```
filteredPostingGroups = append(filteredPostingGroups, postingGroup{isSubtract: false, keys: []labels.Label{allPostingsLabel}})
```

Reply:
Oh you're right! Sorry.

Reply:
this should be very rate I think. We add `allPostings` only if there were no positive matchers. Most queries will have a metric name, which is a positive matcher. https://github.com/grafana/mimir/blob/4effc1f7866a35ec8de4e7cb86dc6ee7dff66f84/pkg/storegateway/bucket_index_reader.go#L293

Reply:
But adding `+1` is basically _free_, while growing the slice in those cases is an extra allocation.

Reply:
If you're really concerned about that `+1` (we shouldn't :D), we can use the loop above:
```
	for _, m := range ms {
		// Each group is separate to tell later what postings are intersecting with what.
		rawPostingGroups = append(rawPostingGroups, toRawPostingGroup(m))
	}
```

To do this logic:

```go
		// If the group is a subtraction group, we must fetch all postings and remove the ones that this matcher selects.
		allRequested = allRequested || pg.isSubtract
		hasAdds = hasAdds || !pg.isSubtract
```

So we'd know whether allRequested before allocating the new slice :)

Reply:
since you say it's cheap, i'll add it. My memory optimizations intuition is still not very advanced